### ISSUE TITLE
Questions related to the experimental setup

### ISSUE THREAD

**Issue by CBY-9527** [3DAgentWorld/S3PO-GS#1]
Thank you for sharing your work. It has been very insightful and helpful to my own research. I have a few questions I would like to ask:

In Table 1, are the reported rendering performances based on the color-refined results?

I downloaded the sequences mentioned in KITTI_sequence_list.txt from the KITTI dataset (https://www.cvlibs.net/datasets/kitti/eval_odometry.php), but I couldn’t find the corresponding ground truth ("gt") files. Could you please advise where to locate them?

In the slam.py file, it seems that args.ns at line 244 and args.sh at line 262 are not defined. 

I look forward to your response. Thank you again!

**Comment by Ysc-shark** [comment]
Thanks for your interest in our work.

Reported values for all 3DGS SLAM methods in Tab.1 are based on color-refined results.

GT files refer to ground truth pose, which can be downloaded from "-   [Download odometry ground truth poses (4 MB)](https://www.cvlibs.net/datasets/kitti/user_login.php)".

In the slam.py file, args.ns means mapping optimization iteration count when the current local BA window is full (follow MonoGS's setting); args.sh means the order of spherical harmonics in 3DGS (but it seems not to work). 

More detailed explanation will be added in the code later.

**Comment by CBY-9527** [comment]
> Thanks for your interest in our work.
> 
> Reported values for all 3DGS SLAM methods in Tab.1 are based on color-refined results.
> 
> GT files refer to ground truth pose, which can be downloaded from "- [Download odometry ground truth poses (4 MB)](https://www.cvlibs.net/datasets/kitti/user_login.php)".
> 
> In the slam.py file, args.ns means mapping optimization iteration count when the current local BA window is full (follow MonoGS's setting); args.sh means the order of spherical harmonics in 3DGS (but it seems not to work).
> 
> More detailed explanation will be added in the code later.



Thank you for your reply. After configuring the environment as instructed, I encountered the following warning when running the code:
"Warning, cannot find cuda-compiled version of RoPE2D, using a slow PyTorch version instead."
I resolved this by modifying the RoPE CUDA kernel compilation step from cd croco/models/curope/ to cd dust3r/croco/models/curope/.

I then tested the code on the KITTI dataset. While the rendering performance is generally consistent with the results reported in the paper, the localization accuracy is not very satisfactory, particularly on sequences 02, 03, 08, and 10.

Do you have any suggestions for reproducing the localization results reported in the paper more faithfully? Thank you in advance—I look forward to your response.

<img width="988" height="192" alt="Image" src="https://github.com/user-attachments/assets/2308a3aa-0be0-4fcc-a33f-23f0590ca290" />

**Comment by Ysc-shark** [comment]
Thank you for your suggestion regarding the warning message.

The per-sequence results on the KITTI dataset, as reported in our paper, are shown below. All results were obtained from actual runs on our server, and we have preserved the full scenes and pose trajectories.

<img width="1151" height="247" alt="Image" src="https://github.com/user-attachments/assets/0a5cf7d6-130f-49f7-8e37-8160a025fa01" />

As shown, S3PO-GS does not perform well on sequences 02, 03, 08, and 10, but the overall performance gap in your results is larger than we expected.

During development, we observed the following issues:

1. The model's performance degrades significantly when the GPU or CPU is under heavy load.
2. Performance is affected by hardware differences.
3. The method shows instability on certain KITTI sequences, occasionally producing extreme values that lower the average. Specifically, pose drift often occurs in the last ~20 frames along straight paths (e.g., elongated or shortened trajectories), while most of the trajectory remains accurate. Since pose evaluation aligns trajectories to GT before computing errors, such local deviations can disproportionately impact the final results. This also explains why rendering appears unaffected, but localization errors are unstable. 

We are actively working on addressing this issue.

To help reproduce the localization results, you may try setting config["Training"]["prune_num"] = 0. I'll also double-check whether any key components or hyperparameters were lost during code cleanup or release.

Lastly, you can use the following script to quickly collect and format the results:
`python cal_result.py --dataname KITTI`

**Comment by CBY-9527** [comment]
> Thank you for your suggestion regarding the warning message.
> 
> The per-sequence results on the KITTI dataset, as reported in our paper, are shown below. All results were obtained from actual runs on our server, and we have preserved the full scenes and pose trajectories.
> 
> <img alt="Image" width="1151" height="247" src="https://private-user-images.githubusercontent.com/85819178/467525398-0a5cf7d6-130f-49f7-8e37-8160a025fa01.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTI4MjkxMDYsIm5iZiI6MTc1MjgyODgwNiwicGF0aCI6Ii84NTgxOTE3OC80Njc1MjUzOTgtMGE1Y2Y3ZDYtMTMwZi00OWY3LThlMzctODE2MGEwMjVmYTAxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzE4VDA4NTMyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM1NTQ4MDNlMTdhNDU2OWRkYTRmZWU1ZGU3OWMzMDEzNzcyYmM5OTVmZWEwOTM0MmZlZjJhZjc4Y2M5ZThkOTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.SrpDpLyu7wZ0rWTbEY8kgHS9OtjQ3qDvHEIeI6oKNJY">
> As shown, S3PO-GS does not perform well on sequences 02, 03, 08, and 10, but the overall performance gap in your results is larger than we expected.
> 
> During development, we observed the following issues:
> 
> 1. The model's performance degrades significantly when the GPU or CPU is under heavy load.
> 2. Performance is affected by hardware differences.
> 3. The method shows instability on certain KITTI sequences, occasionally producing extreme values that lower the average. Specifically, pose drift often occurs in the last ~20 frames along straight paths (e.g., elongated or shortened trajectories), while most of the trajectory remains accurate. Since pose evaluation aligns trajectories to GT before computing errors, such local deviations can disproportionately impact the final results. This also explains why rendering appears unaffected, but localization errors are unstable.
> 
> We are actively working on addressing this issue.
> 
> To help reproduce the localization results, you may try setting config["Training"]["prune_num"] = 0. I'll also double-check whether any key components or hyperparameters were lost during code cleanup or release.
> 
> Lastly, you can use the following script to quickly collect and format the results: `python cal_result.py --dataname KITTI`

Thanks for your prompt reply. I tried setting config["Training"]["prune_num"] = 0, but the RMSE metric did not improve. I examined the generated evo_2dplot, and as you mentioned, most of the trajectory remains accurate, but in the last 20 frames or so, the trajectory tends to be stretched or compressed. While it’s true that hardware can affect performance for SLAM systems with multi-process training, I’m not quite sure why such stretching or compression only occurs at the final stage, whereas the initial RMSE remains normal. Moreover, I checked the input sequence, and in the last few dozen frames, there doesn’t seem to be any sudden change in the camera's motion speed.

**Comment by CBY-9527** [comment]
> Thank you for your suggestion regarding the warning message.
> 
> The per-sequence results on the KITTI dataset, as reported in our paper, are shown below. All results were obtained from actual runs on our server, and we have preserved the full scenes and pose trajectories.
> 
> <img alt="Image" width="1151" height="247" src="https://private-user-images.githubusercontent.com/85819178/467525398-0a5cf7d6-130f-49f7-8e37-8160a025fa01.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTMyNTQzODcsIm5iZiI6MTc1MzI1NDA4NywicGF0aCI6Ii84NTgxOTE3OC80Njc1MjUzOTgtMGE1Y2Y3ZDYtMTMwZi00OWY3LThlMzctODE2MGEwMjVmYTAxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzIzVDA3MDEyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEzZjdiYjE3ZjRhMTA5NmIxOGRmNDMwM2E0NjRmYTVkZTVjZDc1ODRkNmNiYTJmNTFhOTI3YTE0YjFmZWYxNmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.3pEoz4bH9bW3LO_Ulv5ufAyh4I5JeXlDzLJD-EgpVIg">
> As shown, S3PO-GS does not perform well on sequences 02, 03, 08, and 10, but the overall performance gap in your results is larger than we expected.
> 
> During development, we observed the following issues:
> 
> 1. The model's performance degrades significantly when the GPU or CPU is under heavy load.
> 2. Performance is affected by hardware differences.
> 3. The method shows instability on certain KITTI sequences, occasionally producing extreme values that lower the average. Specifically, pose drift often occurs in the last ~20 frames along straight paths (e.g., elongated or shortened trajectories), while most of the trajectory remains accurate. Since pose evaluation aligns trajectories to GT before computing errors, such local deviations can disproportionately impact the final results. This also explains why rendering appears unaffected, but localization errors are unstable.
> 
> We are actively working on addressing this issue.
> 
> To help reproduce the localization results, you may try setting config["Training"]["prune_num"] = 0. I'll also double-check whether any key components or hyperparameters were lost during code cleanup or release.
> 
> Lastly, you can use the following script to quickly collect and format the results: `python cal_result.py --dataname KITTI`




After carefully checking the code and print outputs, I found that the line "if np.any(accurate_pixels) and ((k < 2) or (np.sum(accurate_pixels) >= min_accurate_pixels)):" in the depth_utils.py file might be a potential reason for the trajectory being stretched or shortened. I made a simple modification to this part of the code, and now the performance on various video sequences in the KITTI dataset reaches the values reported in the table you provided. Thank you again for sharing this work and for your previous response.

**Comment by xiaoxiaozlx** [comment]
Hello! I have encountered the same problem. In the test of the straight-line trajectory, there is a relatively serious pose drift phenomenon in the latter part. Is there any solution or idea for this situation at present?

**Comment by CBY-9527** [comment]
您好，我之前仔细检查了下代码，修改了一下depth_utils.py文件中第121行，因为我发现在训练过程中k为0、1时有时出现很少的“准确像素”，这可能会极大干扰尺度因子。初步实验后在kitti上的性能确实不错，有几个序列性能超过给出的表格上的性能。但是，我后来又尝试了几次，发现性能还是不行，性能波动挺大，平均rmse指标在1.5左右。  我有几点思考：1. 方法采用了多进程MonoGS做基线，多进程性能本身就不够稳定，因此会出现不同硬件设备之间存在性能差异；2. depth_utils.py中的process_depth函数中初始迭代时存在准确像素过少情况影响尺度因子计算；3. k为2、3且准确像素过少时会使用find_scale函数，但我发现帧间尺度因子波动很大时往往发生在使用find_scale获得的尺度因子。      如果方便的话，能否添加一下微信以便交流（ycb2468）
  
---- Replied Message ----
From Sicheng Yu ***@***.***> Date 08/06/2025 20:59 To 3DAgentWorld/S3PO-GS ***@***.***> Cc Fisher ***@***.***>,
Author ***@***.***> Subject Re: [3DAgentWorld/S3PO-GS] Questions related to the experimental setup (Issue #1)  
Ysc-shark left a comment (3DAgentWorld/S3PO-GS#1)
Hi,

 Thank you for your comment — I saw that you mentioned making some modifications and successfully reproducing the performance on KITTI.
 However, I couldn't find your new comment under the GitHub issue.

 Just checking: has your issue been resolved?

 Best regards,
 Sicheng Yu



 ysc
 ***@***.***



 &nbsp;




 ------------------&nbsp;原始邮件&nbsp;------------------
 发件人: "3DAgentWorld/S3PO-GS" ***@***.***&gt;;
 发送时间:&nbsp;2025年7月23日(星期三) 下午3:01
 ***@***.***&gt;;
 ***@***.******@***.***&gt;;
 主题:&nbsp;Re: [3DAgentWorld/S3PO-GS] Questions related to the experimental setup (Issue #1)



 CBY-9527 left a comment (3DAgentWorld/S3PO-GS#1)

 Thank you for your suggestion regarding the warning message.

 The per-sequence results on the KITTI dataset, as reported in our paper, are shown below. All results were obtained from actual runs on our server, and we have preserved the full scenes and pose trajectories.
 As shown, S3PO-GS does not perform well on sequences 02, 03, 08, and 10, but the overall performance gap in your results is larger than we expected.
 During development, we observed the following issues:

 The model's performance degrades significantly when the GPU or CPU is under heavy load.

 Performance is affected by hardware differences.

 The method shows instability on certain KITTI sequences, occasionally producing extreme values that lower the average. Specifically, pose drift often occurs in the last ~20 frames along straight paths (e.g., elongated or shortened trajectories), while most of the trajectory remains accurate. Since pose evaluation aligns trajectories to GT before computing errors, such local deviations can disproportionately impact the final results. This also explains why rendering appears unaffected, but localization errors are unstable.

 We are actively working on addressing this issue.

 To help reproduce the localization results, you may try setting config["Training"]["prune_num"] = 0. I'll also double-check whether any key components or hyperparameters were lost during code cleanup or release.

 Lastly, you can use the following script to quickly collect and format the results: python cal_result.py --dataname KITTI

 After carefully checking the code and print outputs, I found that the line "if np.any(accurate_pixels) and ((k < 2) or (np.sum(accurate_pixels) &gt;= min_accurate_pixels)):" in the depth_utils.py file might be a potential reason for the trajectory being stretched or shortened. I made a simple modification to this part of the code, and now the performance on various video sequences in the KITTI dataset reaches the values reported in the table you provided. Thank you again for sharing this work and for your previous response.

**Comment by Ysc-shark** [comment]
随机性很大一部分是来源于**前后端不一致**，这在序列的后期问题会更显著。我们最近发现了一个bug
https://github.com/3DAgentWorld/S3PO-GS/blob/d3a5c0926e5c5bceac362e5d0646e60575cbcd0a/utils/slam_frontend.py#L467
这一行代码导致了前后端不一致，应该被注释。但注释后有些场景反而会变差，目前还在调试。

随机性的另外一部分问题，我认为我们的高斯初始化不合理，采用的是点云随机下采样。
[ARTDECO](https://city-super.github.io/artdeco/)和[on-the-fly 3DGS](https://repo-sam.inria.fr/nerphys/on-the-fly-nvs/)在这点上做的很好。

find_scale是一种弥补策略，并不是很稳定，这点需要改进。

**Comment by peelstnac** [comment]
Hi, great work.

I'm curious why the model's performance degrades significantly when the GPU or CPU is under heavy load?

Thanks!
